gapic/src/main/com/google/gapid/views/StateView.java

```diff
@@ -330,6 +348,67 @@ public void onStateSelected(Path.Any path) {
         });
   }
 
+  private List<Path.Any> getExpandedPaths() {
+    return stream(viewer.getExpandedElements())
+        .map(element -> ((ApiState.Node)element).getData())
+        .filter(data -> data != null)
+        .map(data -> data.getValuePath())
+        .collect(Collectors.toList());
+  }
+
+  protected void updateExpansionState(List<Path.Any> paths, int retry) {
+    Path.State state = Paths.stateAfter(models.state.getSource().getStateTree().getAfter());
+    ApiState.Node root = models.state.getData();
+    List<ListenableFuture<TreePath>> futures = Lists.newArrayList();
+    for (Path.Any path : paths) {
+      Path.Any reparented = Paths.reparent(path, state);
+      if (reparented == null) {
+        LOG.log(WARNING, "Unable to reparent path {0}", path);
+        continue;
+      }
+      futures.add(Futures.transformAsync(models.state.resolve(reparented),
+          nodePath -> getTreePath(root, nodePath)));
+    }
+
+    Rpc.listen(Futures.allAsList(futures),
+        new UiCallback<List<TreePath>, TreePath[]>(viewer.getTree(), LOG) {
+      @Override
+      protected TreePath[] onRpcThread(Result<List<TreePath>> result)
+          throws RpcException, ExecutionException {
+        List<TreePath> list = result.get();
+        return list.toArray(new TreePath[list.size()]);
+      }
+
+      @Override
+      protected void onUiThread(TreePath[] treePaths) {
+        setExpanded(treePaths, paths, retry);
```

Comment:
I think it's due to the dynamic loading and laziness of our tree model that the tree viewer sometimes gives up and doesn't actually expand a path. I detect that by asking if the path is expanded right after expanding it and if it's not expanded, schedule a retry. The retry count is just there to avoid an infinite loop if things go sideways. It does have the additional benefit that if, after playing with Tree (A) and selecting an atom and loading Tree (B), you select another atom quickly - loading Tree (C) - before Tree (B) had a chance to update its expansion state, it will eventually restore Tree (A)'s expansion state, rather than Tree (B)'s empty one.